src/chars.rs

```diff
@@ -39,56 +39,60 @@ pub fn from_byte_idx(text: &str, byte_idx: usize) -> usize {
+        // complexity hurts performance.
+        let mut char_count = 0;
+        for (i, byte) in text.iter().enumerate() {
+            char_count += is_leading_byte(byte);
+            if char_count > char_idx {
+                return i;
+            }
+        }
+        return text.len();
+    }
     // Get `middle` so we can do more efficient chunk-based counting.
     // We can't use this to get `end`, however, because the start index of
     // `end` actually depends on the accumulating char counts during the
     // counting process.
-    let (start, middle, _) = unsafe { text.as_bytes().align_to::<T>() };
+    let (start, middle, _) = unsafe { text.align_to::<T>() };
 
     let mut byte_count = 0;
     let mut char_count = 0;
 
     // Take care of any unaligned bytes at the beginning.
     for byte in start.iter() {
-        char_count += ((*byte & 0xC0) != 0x80) as usize;
+        char_count += is_leading_byte(byte);
         if char_count > char_idx {
-            break;
+            return byte_count;
         }
         byte_count += 1;
     }
 
-    // Process chunks in the fast path.
-    let mut chunks = middle;
-    let mut max_round_len = char_idx.saturating_sub(char_count) / T::MAX_ACC;
-    while max_round_len > 0 && !chunks.is_empty() {
-        // Choose the largest number of chunks we can do this round
-        // that will neither overflow `max_acc` nor blast past the
-        // char we're looking for.
-        let round_len = T::MAX_ACC.min(max_round_len).min(chunks.len());
-        max_round_len -= round_len;
-        let round = &chunks[..round_len];
-        chunks = &chunks[round_len..];
-
-        // Process the chunks in this round.
-        let mut acc = T::zero();
-        for chunk in round.iter() {
-            acc = acc.add(chunk.bitand(T::splat(0xc0)).cmp_eq_byte(0x80));
-        }
-        char_count += (T::SIZE * round_len) - acc.sum_bytes();
-        byte_count += T::SIZE * round_len;
+    // Process chunks in the fast path. Ensure that we don't go past the number
+    // of chars we are counting towards
+    let fast_path_chunks = middle.len().min((char_idx - char_count) / T::SIZE);
+    let bytes = T::SIZE * 4;
+    for chunks in middle[..fast_path_chunks].chunks_exact(4) {
+        let val1 = count_trailing_chunk(chunks[0]);
+        let val2 = count_trailing_chunk(chunks[1]);
+        let val3 = count_trailing_chunk(chunks[2]);
+        let val4 = count_trailing_chunk(chunks[3]);
+        char_count += bytes - val1.add(val2).add(val3.add(val4)).sum_bytes();
+        byte_count += bytes;
     }
 
-    // Process chunks in the slow path.
-    for chunk in chunks.iter() {
-        let new_char_count =
-            char_count + T::SIZE - chunk.bitand(T::splat(0xc0)).cmp_eq_byte(0x80).sum_bytes();
+    // Process the rest of chunks in the slow path.
+    let rest_idx = fast_path_chunks - fast_path_chunks % 4;
+    for chunk in middle[rest_idx..].iter() {
```

Comment:
Rather than a separate `rest_idx` variable, I'd rather just inline the expression since it's only used in one place.